### PR TITLE
Add new tasks create_doi_sunet and contribs

### PR DESCRIPTION
**Pull request by edsu**
Once the initial DOI collection process is complete we know the population of DOIs we are working with in the dataset. We are also able to map the DOI to a SUNETID using either the `orcidid` (for Dimensions and OpenAlex) or `cap_profile_id` (for sul_pub).

The new `doi_sunet` task will create a mapping of `doi -> [sunetid]` using the pickle files, sul_pub csv and the authors csv. This is then used by the `doi_set` task to generate the list of DOIs needed for harvesting.

Once the publications datasets are merged the new `contribs` task uses the `doi_sunet` mapping to add the `sunetid` column, split out the publications into contributions where each row has a unique `sunetid`. Finally the contributions are joined with the `authors.csv`.

### CODE REVIEW

File: rialto_airflow/dags/harvest.py [hunk 1]
```diff
@@ -1,14 +1,14 @@
...
+from airflow.models import Variable
 
-from rialto_airflow.utils import create_snapshot_dir, rialto_authors_file
-from rialto_airflow.harvest import dimensions, openalex, merge_pubs
+from rialto_airflow.harvest import dimensions, merge_pubs, openalex
+from rialto_airflow.harvest.doi_sunet import create_doi_sunet_pickle
 from rialto_airflow.harvest.sul_pub import sul_pub_csv
-from rialto_airflow.harvest.doi_set import create_doi_set
-
+from rialto_airflow.utils import create_snapshot_dir, rialto_authors_file
```
Comment (edsu) [inline rialto_airflow/dags/harvest.py:11]: This was mostly ruff reordering the imports.

File: rialto_airflow/dags/harvest.py [hunk 2]
```diff
@@ -70,12 +70,22 @@ def sul_pub_harvest(snapshot_dir):
         return str(csv_file)
 
     @task()
-    def doi_set(dimensions, openalex, sul_pub):
```
Comment (edsu) [inline rialto_airflow/dags/harvest.py:73]: This is the new task to create the DOI -> [SUNETID] mapping pickle file.

File: rialto_airflow/dags/harvest.py [hunk 3]
```diff
@@ -105,18 +115,12 @@ def merge_publications(sul_pub, openalex_pubs, dimensions_pubs, snapshot_dir):
         return str(output)
 
     @task()
```
Comment (edsu) [inline rialto_airflow/dags/harvest.py:117]: The join to authors is handled as part of pubs_to_contribs now.

File: rialto_airflow/dags/harvest.py [hunk 4]
```diff
@@ -70,12 +70,22 @@ def sul_pub_harvest(snapshot_dir):
...
+
+    @task()
+    def doi_set(doi_sunet_pickle):
         """
-        Extract a unique list of DOIs from the dimensions doi-orcid dict,
-        the openalex doi-orcid dict, and the SUL-Pub publications.
+        Use the DOI -> [SUNET] pickle to return a list of all DOIs.
         """
-        return create_doi_set(dimensions, openalex, sul_pub)
+        return list(pickle.load(open(doi_sunet_pickle, "rb")).keys())
```
Comment (edsu) [inline rialto_airflow/dags/harvest.py:88]: The step of creating the list of DOIs from the `doi_sunet.pickle` is now so simple it's put inline here.

File: rialto_airflow/harvest/doi_set.py
Comment (edsu) [inline rialto_airflow/harvest/doi_set.py:1]: `harvest.doi_set` was renamed to `harvest.doi_sunet`

File: rialto_airflow/harvest/doi_sunet.py
Comment (edsu) [inline rialto_airflow/harvest/doi_sunet.py:1]: This is a complicated bit of code. The `doi_sunet` module contains the logic for creating a `doi -> [sunetid]` mapping by using:

- the `doi -> [orcid]` mappings in the dimensions and openalex pickle files
- the `doi -> [cap_profile_id]` mapping present in the sul_pub.csv

and then generating a `doi -> [sunetid]` by looking up the `sunetid` in the `authors.csv` using the respective `orcid` or `cap_profile_id`.

File: rialto_airflow/harvest/doi_sunet.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+
+def sulpub_doi_sunetids(sul_pub_csv, cap_profile_sunet):
+    # create a dataframe for sul_pubs which has a column for cap_profile_id
+    # extracted from the authorship column
+    df = pd.read_csv(sul_pub_csv, usecols=["doi", "authorship"])
+    df = df[df["doi"].notna()]
+
+    def extract_cap_ids(authors):
+        return [a["cap_profile_id"] for a in eval(authors)]
```
Comment (lwrubel) [inline rialto_airflow/harvest/doi_sunet.py:56]: I think this needs to account for the `status` being `approved` and not using those that have a status of `unknown` or `denied`.

**Review by lwrubel — APPROVED** [review]

**Pull request merged**